meta-lxatac-bsp: barebox: update to 2024.03.0 with Gen 3 support

hnez opened the pull request:
This PR updates the barebox bootloader version on the TAC to version 2024.03.0 and adds support for the most recent hardware generation (Gen 3).

### Related Pull Requests

This PR is based on another PRs that should be merged first:

- [x] The `meta-lxatac` pull request that last touched the barebox patch stack: #94

hnez committed:
meta-lxatac-bsp: barebox: update to 2023.10.0

This barebox version imports our mainline LXA TAC devicetrees from
linux, greatly reducing our local patch stack size.
It also contains some patches that have since been mainlined,
also reducing our patch stack size.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
meta-lxatac-bsp: barebox: update to 2023.12.0

No large and interesting changes, just a normal update to stay up to date.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
meta-lxatac-bsp: barebox: update to config that is actually used

This is just a menuconfig run without changing any options.
The VIDEO_VPL config option is automatically selected by the following
combination of other options:

  DRIVER_VIDEO_PANEL_MIPI_DBI [=y] && VIDEO [=y] && OFTREE [=y] && SPI [=y]

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
meta-lxatac-bsp: barebox: disable boards we do not actually support

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
meta-lxatac-bsp: barebox: disable features we do not need

It is neat to have some debugging features in the bootloader
when you need them.
But we have accumulated a few features that we definitely do
not use.

Disable those.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
meta-lxatac-bsp: barebox: enable fastboot oem command

This allows us to run barebox commands without the need for a serial
console (using the dedicated serial cable or the barebox CDC ACM gadget).

This solves an issue where the TAC would just continue auto booting
when trying to flash a new image onto the TAC via USB.

Also document in the README that we no longer need workarounds for
the watchdog and can now boot the flashed image without requiring a
power cycle.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
meta-lxatac-bsp: barebox: update to 2024.03.0 with Gen 3 support

The main change in this upgrade is the added support for the new generation
of LXA TAC devices (LXA TAC Gen 3).

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
meta-lxatac-bsp: barebox: only interrupt boot on Ctrl+C

Presses of the any key can be triggered by noise on the UART or
button presses on the TAC itself, both of which should not interrupt
the boot process.

Instead only interrupt the boot on Ctrl+C.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez commented:
Hi @a3f, would you be interested in reviewing this pull request and also have a look at the ones leading up to this one?
We would like to release a new meta-lxatac / tacos stable release containing LXA TAC hardware generation 3 support some time this week to get ready to ship devices, so it would be great to get this reviewed sooner than later.

a3f commented on README.md:
```diff
@@ -486,6 +486,11 @@ Next you can upload the required pieces of software:
+    # Start a fastboot command in the background that disables the watchdog
+    # once the LXA TAC shows up as fastboot device.
+    # This also stops the automatic boot process.
+    $ fastboot oem exec "wd -x" &
```
AFAIK, the watchdog is only enabled before booting Linux (`boot` command), so `wd -x` directly after DFU bring-up doesn't matter for watchdogs. As for stopping the automatic boot process, `fastboot flash` has the same effect, see:
https://github.com/barebox/barebox/blob/5423885658da4beddec619c1ec4b1cb741c68554/common/fastboot.c#L814

If this change fixes something, I suspect the issue is elsewhere.

a3f replied:
What I usually do is have an init script:
```
#!/bin/sh

if [ "$bootsource" = serial ]; then
   global.autoboot_timeout=30
fi
```

hnez replied:
Hmm. I would not want to change this back without re-investigating the `wd -x` change again.
But I would also prefer not to do such an investigation before we have our Gen 3 compatible release.
How about a compromise in the form of a draft PR #133 that does the suggested change and has a TODO item to re-investigate `wd -x` in the README?

a3f replied:
Ok.

a3f commented on meta-lxatac-bsp/recipes-bsp/barebox/barebox_2023.10.0.bb:
```diff
@@ -6,10 +6,12 @@ SRC_URI += " \
 
 COMPATIBLE_MACHINE = "lxatac"
 
+BAREBOX_IMAGES = "*.stm32"
```
It would be nice if we could import the [barebox recipe that eventually should make it into OE-core](https://lore.kernel.org/all/20230425184720.456896-1-ejo@pengutronix.de/) in a future update. This parses the `barebox-flash-images` file, so hardcoding `BAREBOX_IMAGES` as done here is unneeded.

hnez replied:
I've done so now in a separate commit. Please have a look at it.

a3f replied:
LGTM. Thanks for giving this some wider testing!

a3f commented on meta-lxatac-bsp/recipes-core/images/emmc-boot-image/lxatac/genimage.config:
```diff
@@ -18,7 +18,7 @@ image @IMAGE@fip {
         fw-config = "stm32mp157c-lxa-tac-fw-config.dtb"
         hw-config = "stm32mp157c-lxa-tac-fw-config.dtb"
-        nt-fw = "barebox-stm32mp15xc-lxa-tac.img"
+        nt-fw = "barebox-stm32mp15xc-lxa-tac.stm32.img"
```
I think this should go away with the to-be-upstreamed OE-core recipe.

hnez replied:
It just spins the filename extension carousel one step further by changing the extension to `barebox-stm32mp15xc-lxa-tac.stm32`.

a3f replied:
that's the correct name though. `.img` is misleading as the `.stm32` images only work by chance for your use case: The stm32 header in front is no longer parsed by TF-A and is directly executed. I think we are lucky that the header evaluates to some arithmetic instructions that don't crash.

hnez replied:
Huh, I don't like the sound of that. What would be the proper solution here?
Is it something we should squeeze in with this PR or address in an issue?

a3f replied:
I can submti a follow-up PR if you test it.

hnez replied:
deal!

a3f replied:
I couldn't rebuild the umpf tag, so here goes my change. Please test and squash if it works:
```
commit 4d6c3ef497f7d39c15cbd3923a1c038026c2653a (HEAD)
Author:     Ahmad Fatoum <a.fatoum@pengutronix.de>
AuthorDate: Thu Apr 4 22:16:35 2024 +0200
Commit:     Ahmad Fatoum <a.fatoum@pengutronix.de>
CommitDate: Thu Apr 4 22:18:09 2024 +0200

    fixup! ARM: stm32mp: add Linux Automation TAC board
    
    ARM: stm32mp: lxa-tac: remove STM32 header from bootloader binary
    
    The STM32 header is what's parsed by the BootROM of the STM32MP1. TF-A
    used to expect the same header for the BL33 binary it loads, but it no
    longer does and instead expects a FIP image that contains amongs others
    the BL33 binary as directly executable binary without header and a
    device tree that it passes to it.
    
    For use with FIP, barebox provides a barebox-stm32mp-generic-bl33.img
    image that is not board-specific. We don't use that for the LXA TAC as
    we don't want to settle on a single device tree and instead want to
    select the correct device tree at runtime from multiple that are linked
    into barebox.
    
    Instead we continue to use the .stm32 image, which still has the STM32
    header, which is now directly executed. This didn't lead to problems so
    far, because the header happens to translate to valid ARM arithmetic
    instructions that don't break anything, but we shouldn't rely on that.
    
    Fix this by generating a plain image instead without header.
    
    Signed-off-by: Ahmad Fatoum <a.fatoum@pengutronix.de>

diff --git a/images/Makefile.stm32mp b/images/Makefile.stm32mp
index e8c7adc9d8f6..80f6b9dfc8f5 100644
--- a/images/Makefile.stm32mp
+++ b/images/Makefile.stm32mp
@@ -34,7 +34,10 @@ $(call build_stm32mp_image, CONFIG_MACH_STM32MP15XX_DKX, start_stm32mp15xx_dkx,
 $(call build_stm32mp_image, CONFIG_MACH_STM32MP15X_EV1, start_stm32mp15x_ev1, stm32mp15x-ev1)
 
 $(call build_stm32mp_image, CONFIG_MACH_LXA_MC1, start_stm32mp157c_lxa_mc1, stm32mp157c-lxa-mc1)
-$(call build_stm32mp_image, CONFIG_MACH_LXA_TAC, start_stm32mp15xc_lxa_tac, stm32mp15xc-lxa-tac)
+
+pblb-$(CONFIG_MACH_LXA_TAC) += start_stm32mp15xc_lxa_tac
+FILE_barebox-stm32mp15xc-lxa-tac.img  = start_stm32mp15xc_lxa_tac.pblb
+image-$(CONFIG_MACH_LXA_TAC) += barebox-stm32mp15xc-lxa-tac.img
 
 $(call build_stm32mp_image, CONFIG_MACH_PROTONIC_STM32MP1, start_prtt1a, prtt1a)
 $(call build_stm32mp_image, CONFIG_MACH_PROTONIC_STM32MP1, start_prtt1s, prtt1s)
```

hnez replied:
I've applied the change and tested it now and it seems to work fine.

a3f replied:
I guess that means this PR is ok to merge then.

a3f commented on meta-lxatac-bsp/recipes-bsp/barebox/files/patches/0102-ARM-stm32mp-add-Linux-Automation-TAC-board.patch:
```diff
@@ -0,0 +1,421 @@
++#include <arm/st/stm32mp157c-lxa-tac-gen2.dts>
++#include "stm32mp15xc-lxa-tac.dtsi"
++
++/ {
```
Please cherry-pick this fixup:
```
commit 840911ef51ae1e6a8b477ea5c9e96dcbd2c9f9e2 (downstream/v2023.10.0/customers/lxa/tac)
Author:     Ahmad Fatoum <a.fatoum@pengutronix.de>
AuthorDate: Mon Mar 18 08:01:08 2024 +0100
Commit:     Ahmad Fatoum <a.fatoum@pengutronix.de>
CommitDate: Mon Mar 18 08:02:59 2024 +0100

    fixup! ARM: stm32mp: add Linux Automation TAC board
    
    There's WIP-support for the other binding in
    v2024.03.0/topic/gpio-pull-up, but until that's cleaned up and upstream,
    just add a local workaround.
    
    Reported-by: Chris Fiege <c.fiege@pengutronix.de>
    Signed-off-by: Ahmad Fatoum <a.fatoum@pengutronix.de>

diff --git a/arch/arm/dts/stm32mp15xc-lxa-tac.dtsi b/arch/arm/dts/stm32mp15xc-lxa-tac.dtsi
index 33f83e035457..00ba39a7a5be 100644
--- a/arch/arm/dts/stm32mp15xc-lxa-tac.dtsi
+++ b/arch/arm/dts/stm32mp15xc-lxa-tac.dtsi
@@ -91,6 +91,43 @@ last_chosen@10 {
 	};
 };
 
+/*
+ * There are two bindings for pin bias:
+
+ * - The pinctrl binding, which differs from controller to controller,
+ *   but is capable of representing controller specific attributes like
+ *   pull ups of varying strength. This is used via
+ *   pinctrl-names/pinctrl-[0-9]+ DT properties
+ *
+ * - Or-ing GPIO_PULL_UP or GPIO_PULL_DOWN into the GPIO flags after the
+ *   phandle, e.g.  gpios = <&gpioe 7 (GPIO_ACTIVE_LOW | GPIO_PULL_UP)>;
+ *   Thix can expose a generic interface, but doesn't allow fine tuning.
+ *
+ * Not all Linux drivers support the second binding and barebox doesn't
+ * support it all; It requires pinctrl drivers to expose to the core
+ * a way to configure pull ups/pull downs, but we don't export this info.
+ * Instead drivers are called with the device tree node in action and
+ * instructed to apply it.
+ *
+ * We can add support for this binding for STM32 only as first step, but
+ * till then, let's just use the first binding instead to configure the
+ * GPIO pull-ups.
+ */
+&{/gpio-keys} {
+	pinctrl-names = "default";
+	pinctrl-0 = <&lxatac_gpio_key_pins>;
+};
+
+&pinctrl {
+	lxatac_gpio_key_pins: lxatac-gpio-key-pins {
+		pins {
+			pinmux = <STM32_PINMUX('E',  7, GPIO)>, /* USER_BTN2 */
+				<STM32_PINMUX('I', 11, GPIO)>; /* USER_BTN */
+			bias-pull-up;
+		};
+	};
+};
+
 &baseboard_eeprom {
 	#address-cells = <1>;
 	#size-cells = <1>;
```

hnez replied:
done

a3f commented on meta-lxatac-bsp/recipes-bsp/barebox/files/lxatac/env/nv/autoboot_abort_key:
```diff
@@ -0,0 +1 @@
+ctrl-c
```
The noise on the UART should be resolved by correct the pull settings in the fixup mentioned earlier.

hnez replied:
The UART should be unaffected, shouldn't it? It's the button pullups that have changed.
I think we could keep this change. It should make the boot more robust to all kinds of interference.

a3f replied:
Yes, you're right.

hnez committed:
meta-lxatac-bsp: barebox: add patch that enables pull-ups on buttons

Previously barebox did not enable pull-up resistors for the buttons on
the TAC, leaving the respective pins floating instead.
This resulted in noise generating a storm of button press events,
which did e.g. mess with the console because one of the buttons deletes
text from the commandline input.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
meta-lxatac-bsp: barebox: re-use barebox.inc that will go into oe-core

This prepares us for a future where oe-core contains a barebox recipe
and where we can use the barebox.inc from there instead of the one from
meta-ptx.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
meta-lxatac-bsp: barebox: build binaries without STM32 header

Ahmad Fatoum has pointed out that the .stm32 images we currently use only
work by chance for your use case because the stm32 header in front is no
longer parsed by TF-A and is directly executed.

It seems that we were just lucky until that the bytes in the header
evaluates to arithmetic instructions that don't crash.

Apply a patch provided by Ahmad that builds the barebox binary without
STM32 header.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>
